test/common/config/registry_test.cc

```diff
@@ -182,6 +194,48 @@ TEST(RegistryTest, TestDoubleRegistrationByName) {
                             "Double registration for name: 'testing.published.test'");
 }
 
+class PublishedFactoryWithNameAndCategory : public PublishedFactory {
+public:
+  std::string category() const override { return "testing.published.additional.category"; }
+  std::string name() const override {
+    return "testing.published.versioned.instead_name_and_category";
+  }
+};
+
+TEST(RegistryTest, DEPRECATED_FEATURE_TEST(VersionedWithDeprecatedNamesFactoryAndNewCategory)) {
+  PublishedFactoryWithNameAndCategory test;
+
+  // Check the category is not registered
+  ASSERT_FALSE(Registry::FactoryCategoryRegistry::isRegistered(test.category()));
+
+  auto factory = Registry::RegisterFactory<PublishedFactoryWithNameAndCategory, PublishedFactory>(
+      FACTORY_VERSION(0, 0, 1, {{"build.kind", "private"}}),
+      {"testing.published.versioned.deprecated_name_and_category"});
+
+  // Check the category now registered
+  ASSERT_TRUE(Registry::FactoryCategoryRegistry::isRegistered(test.category()));
+
+  const auto& factories = Envoy::Registry::FactoryCategoryRegistry::registeredFactories();
+
+  auto version =
+      factories.find("testing.published.additional.category")
+          ->second->getFactoryVersion("testing.published.versioned.instead_name_and_category");
+
+  ASSERT_TRUE(version.has_value());
+  EXPECT_EQ(0, version.value().version().major_number());
+  EXPECT_EQ(0, version.value().version().minor_number());
+  EXPECT_EQ(1, version.value().version().patch());
+  EXPECT_EQ(1, version.value().metadata().fields().size());
+  EXPECT_EQ("private", version.value().metadata().fields().at("build.kind").string_value());
+
+  // Get the version using deprecated name and check that it matches the
+  // version obtained through the new name.
+  auto deprecated_version =
+      factories.find("testing.published.additional.category")
+          ->second->getFactoryVersion("testing.published.versioned.deprecated_name_and_category");
+  EXPECT_THAT(deprecated_version, testing::Optional(ProtoEq(version.value())));
```

Comment:
Minor nit: add a `using testing::Optional` declaration at the top of the enclosing namespace so we don't need the extra `testing::` qualifier here.